### PR TITLE
[node-red] New definitions (rewrites defs for 1 module, adds defs for 5 dep modules)

### PR DESCRIPTION
**Pull request by alexk111**
Please fill in this template.

- [x] Use a meaningful title for the pull request. Include the name of the package modified.
- [x] Test the change in your own code. (Compile and run.)
- [x] Add or edit tests to reflect the change. (Run with `npm test`.)
- [x] Follow the advice from the [readme](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#make-a-pull-request).
- [x] Avoid [common mistakes](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#common-mistakes).
- [x] Run `npm run lint package-name` (or `tsc` if no `tslint.json` is present).

Select one of these and delete the others:

If adding a new definition:
- [x] The package does not already provide its own types, or cannot have its `.d.ts` files generated via `--declaration`
- [x] If this is for an NPM package, match the name. If not, do not conflict with the name of an NPM package.
- [x] Create it with `dts-gen --dt`, not by basing it on an existing project.
- [x] Represents shape of module/library [correctly](https://www.typescriptlang.org/docs/handbook/declaration-files/library-structures.html)
- [x] `tslint.json` should be present and it shouldn't have any additional or disabling of rules. Just content as `{ "extends": "dtslint/dt.json" }`. If for reason the some rule need to be disabled, disable it for that line using `// tslint:disable-next-line [ruleName]`  and not for whole package so that the need for disabling can be reviewed.
- [x] `tsconfig.json` should have `noImplicitAny`, `noImplicitThis`, `strictNullChecks`, and `strictFunctionTypes` set to `true`.

### CODE REVIEW

**Comment by andersea** [comment]
@alexk111 Thanks for the PR. This is a large amount of changes. It almost looks like a complete rewrite. To be honest I don't think I will have time to review it, also I see that you have removed me as a contributor, so you will probably be on your own with this from now on. I am happy to approve the PR with this in mind.

There is the issue with CI breakage. I hope you know what you are doing. :)

**Comment by alexk111** [comment]
@andersea thanks for the feedback!

> It almost looks like a complete rewrite. 

Yes, this is a complete rewrite :) Had to do that, because the current types didn't work well for deeper integrations with node-red.

> I see that you have removed me as a contributor

I removed contributors as I wasn't sure that you'd be willing to review changes in types covering the whole node-red app.

> There is the issue with CI breakage. 

Will be fixed as soon as this dependency would be whitelisted: https://github.com/microsoft/DefinitelyTyped-tools/pull/72

File: types/node-red__util/package.json
Comment (peterblazejewicz) [inline types/node-red__util/package.json:3]: Can you add this entry back? `jsonata`. Without that the PR cannot be verified locally (I understeand it will fail with whitelist requirement)

File: types/node-red/index.d.ts
Comment (peterblazejewicz) [inline types/node-red/index.d.ts:5]: can we add those owners back?

**Comment by peterblazejewicz** [comment]
@alexk111 can you consider to create backward compatible version (v0)?
Here is a sample on how to do this:
https://github.com/DefinitelyTyped/DefinitelyTyped/tree/master/types/color
This is because I consider `node-red` a quite popular package, so there could be a requirement to make changes to v0 in future.

File: types/node-red__util/package.json
Comment (alexk111) [reply types/node-red__util/package.json:3]: Thanks for the heads up! Forgot to add it back after running local tests. Fixed now.

File: types/node-red/index.d.ts
Comment (alexk111) [reply types/node-red/index.d.ts]: Makes sense as the previous version will be kept supported. Done.

**Comment by alexk111** [comment]
@peterblazejewicz I agree that the previous version of types should be kept supported, to not break all the things depending on it. Thanks for the link to the example. Didn't know about the `v` folders standard. Done, the v0 is in the PR.

File: types/node-red/index.d.ts
Comment (peterblazejewicz) [inline types/node-red/index.d.ts:10]:
```suggestion
// tslint:disable:no-empty-interface
```
can we remove this exclusion and remove the namespaces with same name? I understand `namespace X {}, namespace X.A {}` etc, but what's the purpose of those three blocks? IMO it just make things less readable (the outcome is same)

File: types/node-red__editor-client/node-red__editor-client-tests.ts
Comment (peterblazejewicz) [inline types/node-red__editor-client/node-red__editor-client-tests.ts:2]:
```suggestion
```

File: types/node-red__editor-client/node-red__editor-client-tests.ts
Comment (peterblazejewicz) [inline types/node-red__editor-client/node-red__editor-client-tests.ts:94]:
```suggestion
            : function() {
```

File: types/node-red__editor-client/node-red__editor-client-tests.ts
Comment (peterblazejewicz) [inline types/node-red__editor-client/node-red__editor-client-tests.ts:117]:
```suggestion
            : function() {
```

File: types/node-red__editor-client/node-red__editor-client-tests.ts
Comment (peterblazejewicz) [inline types/node-red__editor-client/node-red__editor-client-tests.ts:106]:
```suggestion
            : function() {
```

File: types/node-red__editor-client/node-red__editor-client-tests.ts
Comment (peterblazejewicz) [inline types/node-red__editor-client/node-red__editor-client-tests.ts:190]:
```suggestion
            : function(idx) {
```

File: types/node-red__editor-client/node-red__editor-client-tests.ts
Comment (peterblazejewicz) [inline types/node-red__editor-client/node-red__editor-client-tests.ts:204]:
```suggestion
            : function() {
```

File: types/node-red/index.d.ts
Comment (alexk111) [reply types/node-red/index.d.ts]: Fixed. That was done to visually separate three different groups of interfaces and keep them in a single namespace at the same time. Emphasized with comments instead of the namespace duplicates.

File: types/node-red__editor-client/node-red__editor-client-tests.ts
Comment (alexk111) [reply types/node-red__editor-client/node-red__editor-client-tests.ts:2]: This can't be removed, because the parens are added by prettier. Try to update the `function ()` -> `function()` locally and run `npm run prettier types/node-red__editor-client/node-red__editor-client-tests.ts`.

Recent changes in Prettier 2.0:

> Previously, a space would be added after the function keyword in function declarations, but not in function expressions. Now, for consistency, a space is always added after the function keyword. The only exception is generator declarations where function* is treated as a whole word.

Source: https://prettier.io/blog/2020/03/21/2.0.0.html#always-add-a-space-after-the-function-keyword-3903httpsgithubcomprettierprettierpull3903-by-j-f1httpsgithubcomj-f1-josephfrazierhttpsgithubcomjosephfrazier-sosukesuzukihttpsgithubcomsosukesuzuki-thorn0httpsgithubcomthorn0-7516httpsgithubcomprettierprettierpull7516-by-bakkothttpsgithubcombakkot

**Review by peterblazejewicz — APPROVED** [review]
LGTM!
@alexk111 thx for the contribution!
Waiting for the upstream merge, thx!

**Comment by alexk111** [comment]
@peterblazejewicz Thanks for reviewing it!

**Comment by peterblazejewicz** [comment]
btw: i've created a request to review a `dtslint` rule change for this usecase with `prettier`:
https://github.com/microsoft/dtslint/issues/304

**Comment by andersea** [comment]
@peterblazejewicz & @alexk111 thanks a lot!

**Comment by alexk111** [comment]
@peterblazejewicz ok, cool. i've subscribed to the request, will remove the disable rule when it becomes possible.

**Comment by alexk111** [comment]
https://github.com/microsoft/DefinitelyTyped-tools/pull/72 has been merged

**Comment by peterblazejewicz** [comment]
@alexk111 
Could you amend latest commit and force push changes to upstream? This should trigger CI build with current, updated state.
```bash
git commit --amend --no-edit
git push -f
```
(note: it works on my machine and my git)

**Comment by alexk111** [comment]
@peterblazejewicz Done. That triggered CI build, but still returns the `Error: In package.json: Dependency jsonata not in the allowed dependencies list.` Double-checked the https://github.com/microsoft/DefinitelyTyped-tools/pull/72 - it was merged, but not released. Is this the causes of the issue?

**Comment by peterblazejewicz** [comment]
So it seems it was merged, but didn't get into 0.0.42 release:

**Comment by alexk111** [comment]
Ok. Waiting for the next iteration then.

**Comment by danger-public** [comment]
Inspecting the JavaScript source for this package found some properties that are not in the .d.ts files.
The check for missing properties isn't always right, so take this list as advice, not a requirement.
## node-red__registry ([<kbd>unpkg</kbd>](https://unpkg.com/browse/node-red__registry@latest/))
was missing the following properties:
1. init
1. load
1. clear
1. registerType
1. get

<details>
<summary>as well as these 19 other properties...</summary>
<p>getNodeInfo, getNodeList, getModuleInfo, getModuleList, getNodeConfigs, getNodeConfig, getNodeIconPath, getNodeIcons, enableNode, disableNode, addModule, removeModule, installModule, uninstallModule, cleanModuleList, paletteEditorEnabled, getNodeExampleFlows, getNodeExampleFlowPath, deprecated</p>
</details>

<p align="right">
  Generated by :no_entry_sign: <a href="https://danger.systems/js">dangerJS</a> against b61ca0f6f11d9312e93c01bf23cddd4d7e5ecbba
</p>

**Comment by alexk111** [comment]
@peterblazejewicz Had to add another `// tslint:disable:space-before-function-paren` to the v0 due to a space added by prettier.

**Review by peterblazejewicz — APPROVED** [review]
LGTM!
@alexk111 thx for the contribution!

**Comment by alexk111** [comment]
@peterblazejewicz Going to add a typedef for another module (node-red-node-test-helper), which depends on this one. As this module isn't merged yet, I'm not sure what's the workflow in such cases? Should I create a fork from this module's branch and then mark somehow that the other module should be merged after this one?

**Comment by peterblazejewicz** [comment]
I'd advise to wait with this PR in current state (all green). It requires just a DT maintainer to approve as this one deals with multiple packages changes.
I'd start locally with this PR branch as a parent, but once that PR is merged, I'd reparent onto master:
https://git-scm.com/book/en/v2/Git-Branching-Rebasing ('more interesting rebases`, a.k.a. reparent)

**Comment by alexk111** [comment]
@peterblazejewicz Ok. I'll wait then. Thx!

**Review by andrewbranch — APPROVED** [review]
There’s no way I can review the content of this, but since at least one owner has no objections, there’s a backward-compatible v0, and @peterblazejewicz signed off, I think we can move forward. Thanks!

**Comment by alexk111** [comment]
Ready to merge

**Comment by Shaquu** [comment]
With latest version of types I have some code issues:
RED is NodeRedApp
RED.nodes.registerType
Property 'registerType' does not exist on type 'InternalNodesModule'.

Also:
RED.settings.userDir
Property 'userDir' does not exist on type 'PersistentSettings'.

@alexk111 could you help?

**Comment by alexk111** [comment]
@Shaquu Check the example of how to use the new types in nodes:
Editor side (.html): https://github.com/alexk111/node-red-node-typescript-starter/blob/master/src/nodes/transform-text/transform-text.html/index.ts
Runtime side (.js): https://github.com/alexk111/node-red-node-typescript-starter/blob/master/src/nodes/transform-text/transform-text.ts

Hope this will help.

**Comment by Shaquu** [comment]
@alexk111 thanks but in the seconds example RED is untyped from what I see.

**Comment by alexk111** [comment]
@Shaquu

> in the seconds example RED is untyped

 It's not untyped. It gets the type from "contextual typing" (https://www.typescriptlang.org/docs/handbook/functions.html#inferring-the-types).

**Comment by Shaquu** [comment]
Yes, thank you.

**Pull request merged**
